File: app/src/main/java/com/kontranik/koreader/ImageUtils.java
package com.kontranik.koreader;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import java.io.ByteArrayOutputStream;

public class ImageUtils {

    public static Bitmap drawableToBitmap (Drawable drawable) {

        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable)drawable).getBitmap();
        }

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    public static Bitmap getBitmap(Context c, ImageEnum imageEnum, String path) {
        byte[] b = getBitmapData(c, imageEnum, path );
        if ( b != null) {
            Bitmap bmp = BitmapFactory.decodeByteArray(b, 0, b.length);

            int width = 100;
            int height = 100;

            Bitmap background = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            float originalWidth = bmp.getWidth();
            float originalHeight = bmp.getHeight();

            Canvas canvas = new Canvas(background);

            float scale = width / originalWidth;

            float xTranslation = 0.0f;
            float yTranslation = (height - originalHeight * scale) / 2.0f;

            Matrix transformation = new Matrix();
            transformation.postTranslate(xTranslation, yTranslation);
            transformation.preScale(scale, scale);

            Paint paint = new Paint();
            paint.setFilterBitmap(true);

            canvas.drawBitmap(bmp, transformation, paint);

            return background;
        }
        return  null;
    }

    private static byte[] getBitmapData(Context c, ImageEnum imageEnum, String path) {
        Drawable d = null;
        byte[] cover = null;
        switch (imageEnum) {
            case Parent:
                d = c.getResources().getDrawable(R.drawable.ic_arrow_back_black_24dp);
                break;
            case SD:
                d = c.getResources().getDrawable(R.drawable.ic_sd_card_black_24dp);
                break;
            case Dir:
                d = c.getResources().getDrawable(R.drawable.ic_folder_black_24dp);
                break;
            case Epub:
                cover = EpubHelper.getCover(path);
                break;
            default:
                d = c.getResources().getDrawable(R.drawable.ic_book_black_24dp);
                break;
        }

        if ( d == null && cover == null) {
            d = c.getResources().getDrawable(R.drawable.ic_book_black_24dp);
        }
        if ( d != null ) {
            Bitmap icon = ImageUtils.drawableToBitmap(d);
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            if (icon != null) {
                icon.compress(Bitmap.CompressFormat.PNG, 100, stream);
                cover = stream.toByteArray();
            }
        }

        return cover;
    }


}

File: app/src/main/java/com/kontranik/koreader/utils/ImageUtils.kt
package com.kontranik.koreader.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.Rect
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import androidx.core.content.res.ResourcesCompat
import com.kontranik.koreader.R
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer


object ImageUtils {
    fun drawableToBitmap(drawable: Drawable): Bitmap {
        if (drawable is BitmapDrawable) {
            if (drawable.bitmap != null) {
                return drawable.bitmap
            }
        }

        val bitmap = if (drawable.intrinsicWidth <= 0 || drawable.intrinsicHeight <= 0) {
            Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888) // Single color bitmap will be created of 1x1 pixel
        } else {
            Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
        }

        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    @JvmStatic
    fun getBitmap(c: Context, imageEnum: ImageEnum): Bitmap {

        val d = when (imageEnum) {
            ImageEnum.Parent -> ResourcesCompat.getDrawable(
                c.resources,
                R.drawable.ic_baseline_arrow_back_24,
                null
            )
            ImageEnum.SD -> ResourcesCompat.getDrawable(
                c.resources,
                R.drawable.ic_baseline_sd_card_24,
                null
            )
            ImageEnum.Dir -> ResourcesCompat.getDrawable(
                c.resources,
                R.drawable.ic_folder_black_24dp,
                null
            )
            else -> ResourcesCompat.getDrawable(c.resources, R.drawable.ic_book_black_24dp, null)
        }

        return drawableToBitmap(d!!)
    }

    @JvmStatic
    fun invertAndTint(src: Bitmap, tintColor: Int?): Bitmap? {
        val height = src.height
        val width = src.width
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val paint = Paint()
        val matrixGrayscale = ColorMatrix()
        matrixGrayscale.setSaturation(0f)
        val matrixInvert = ColorMatrix()
        matrixInvert.set(
            floatArrayOf(
                -1.0f, 0.0f, 0.0f, 0.0f, 255.0f,
                0.0f, -1.0f, 0.0f, 0.0f, 255.0f,
                0.0f, 0.0f, -1.0f, 0.0f, 255.0f,
                0.0f, 0.0f, 0.0f, 1.0f, 0.0f
            )
        )
        matrixInvert.preConcat(matrixGrayscale)
        val filter = ColorMatrixColorFilter(matrixInvert)
        paint.colorFilter = filter
        canvas.drawBitmap(src, 0f, 0f, paint)
        tintColor?.let { canvas.drawColor(it, PorterDuff.Mode.MULTIPLY) }
        return bitmap
    }

    fun byteArrayToScaledBitmap(byteArray: ByteArray?, width: Int, height: Int): Bitmap? {
        if ( byteArray != null) {
            val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
            if (bitmap != null)
                return scaleBitmap(bitmap, width, height)
        }
        return null
    }

    fun byteArrayToBitmap(byteArray: ByteArray): Bitmap {
        return BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
    }

    fun scaleBitmap(bitmap: Bitmap, maxWidth: Int, maxHeight: Int): Bitmap {
        val mSize = getScaledSize(bitmap.width, bitmap.height, maxWidth, maxHeight)
        return Bitmap.createScaledBitmap(bitmap, mSize.width(), mSize.height(), true)
    }

    fun getScaledSize(originalWidth: Int, originalHeight: Int, maxWidth: Int, maxHeight: Int): Rect {
        var width: Float = originalWidth.toFloat()
        var height: Float = originalHeight.toFloat()
        if (width > height) {
            // landscape
            val ratio = width / maxWidth
            width = maxWidth.toFloat()
            height /= ratio
        } else if (height > width) {
            // portrait
            val ratio = height / maxHeight
            height = maxHeight.toFloat()
            width /= ratio
        } else {
            // square
            height = maxHeight.toFloat()
            width = maxWidth.toFloat()
        }

        if ( width > maxWidth) {
            val ratio = width / maxWidth
            width = maxWidth.toFloat()
            height /= ratio
        }
        if ( height > maxHeight) {
            val ratio = height / maxHeight
            height = maxHeight.toFloat()
            width /= ratio
        }

        return Rect(0, 0, width.toInt(), height.toInt())
    }

    // convert from bitmap to byte array
    fun getBytes(bitmap: Bitmap?): ByteArray? {
        if (bitmap == null) return null

        val scaled = if (bitmap.width > 50 || bitmap.height > 100) scaleBitmap(bitmap, 50, 100) else bitmap
        val stream = ByteArrayOutputStream()
        scaled.compress(CompressFormat.PNG, 0, stream)
        return stream.toByteArray()
    }

    // convert from byte array to bitmap
    fun getImage(byteArray: ByteArray): Bitmap? {
        return BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
    }
}